marbl/marbl_test.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2015 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package marbl
+
+import (
+	"bytes"
+	"net/http"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/google/martian/proxyutil"
+)
+
+func deserializeHeaders(b bytes.Buffer) map[string]string {
+	headers := make(map[string]string)
+
+	reader := NewReader(&b)
+	frame, _ := reader.ReadFrame()
+	for frame != nil {
+		headerFrame, _ := frame.(Header)
+		headers[headerFrame.Name] = headerFrame.Value
+		frame, _ = reader.ReadFrame()
+	}
+
+	return headers
+}
+
+func TestSendTimestampWithLogRequest(t *testing.T) {
+	req, err := http.NewRequest("POST", "http://example.com", nil)
+	if err != nil {
+		t.Fatalf("http.NewRequest(): got %v, want no error", err)
+	}
+	var b bytes.Buffer
+	s := NewStream(&b)
```

Comment:
[marbl.NewStream loops](https://github.com/google/martian/blob/latency/marbl/marbl.go#L59) in a Go routine over the frames as it logs, which introduces a race condition.  I put `time.Sleep(1 * time.Millisecond)` as the first line in [Stream.loop](https://github.com/google/martian/blob/latency/marbl/marbl.go#L64) and the test fails intermittently.

You could introduce a channel that you write to as you call ReadFrame from it's own Go routine, and then read off of it so that you're synchronized after LogRequest.